### ISSUE TITLE
Schedule a one-off for GraphiQL discussions

### ISSUE THREAD

**Issue by orta** [graphql/graphql-wg#136]
👋 - I kickstarted some discussion on twitter about a what could GraphiQL 2.0 (or, I guess 1.0 as it's still not there yet :D ) after exploring some wireframes. 

We should schedule a one-off meeting to talk about:

- What are the current issues with GraphiQL?
- What could / should GraphiQL do?
- Could we find a way to integrate ( or support the integration of) the features from GraphQL playground and/or explorer?

**Comment by benjie** [comment]
I'd also like to give a shout out to GraphiQL Explorer from OneGraph: https://github.com/OneGraph/graphiql-explorer - PostGraphile users have been really enjoying it. We also do some "state hackery" in PostGraphile to enable hot schema replacement without losing your state in the documentation explorer - definitely interested in making this part of core rather than hacking internal React state! Also interested in [fixing some bugs](https://github.com/graphql/graphiql/pull/747), improving the panels situation (as @orta's mockups do), enabling replacement of some of the built-in commands (e.g. using `prettier` rather than `graphql.print` for prettification *on keyboard shortcut* and maintaining cursor position), and much more relating to customisability and extensibility.

Basically: I would like to help improve GraphiQL and I would be happy to help with maintaining the project.

**Comment by sgrove** [comment]
Making GraphiQL much more pluggable would also be nice. I think this would involve:

1. Some refactoring of of the built in components (to make controlling the various pieces easier from the outside)
2. The styling system (to make theming and styling easier)
 - [ ] Along these lines, I believe @sachag has batted around the idea of giving GraphiQL a visual overhaul for the default out-of-the-box-experience, and has already [offered up a logo suggestion](https://github.com/graphql/graphiql/issues/763). It'd be great to see GraphiQL receive not just technical, but visual and UX love as well from the community.
3. Adding some more built-in components that could be (optionally) used just like the existing components:
 - [ ] A header control tab
 - [ ] Tabs for multiple queries
 - [ ] A form-based view for the variable inputs (with the exception of custom scalars, we can reflect enough on the variable inputs to generate decent forms)
 - [ ] Multi-operation support for variable inputs (right now there's one variable input field, and it gets sent to whatever operation is executed, but this makes it a bit hard to test different operations successively)
4. More plugin/extension support. We have a few plugins, like the [GraphiQL Explorer](
https://github.com/OneGraph/graphiql-explorer) (as @benjie kindly pointed out) that fit in reasonably well as-is, but it might be nice to make this more straight-forward. We have some more plugins in mind:
 - [ ] A "query doctor" that help guide the user to fix up their queries (Adding names to operations, extracting variables from hard-coded values, noting potential fragment uses, etc.). The goal is to gently nudge the user towards best practices as though they had an experience graphql user next to them.
 - [ ] A code exporter (ultimately to be based on the amazing work by https://graphql-code-generator.com) that makes getting queries *out* of GraphiQL and *into* a practical context a copy/paste job
 - [ ] The variable-input-as-a-form idea above

So in summary, I think most of the above points fall into:
1. Better styling support
2. More built-in (but opt-in) components
3. More clear plugin/extension points

**Comment by SachaG** [comment]
Re: UI overhaul, here's the new Storybook design by @domyen as inspiration:

![https://cdn-images-1.medium.com/max/2000/1*SwnoxDnjVXKzCVQbD2mHQw.png](https://cdn-images-1.medium.com/max/2000/1*SwnoxDnjVXKzCVQbD2mHQw.png)

**Comment by domyen** [comment]
This is an awesome initiative! I'm pretty familiar with old GraphiQL from my Apollo days. It sounds like the new GraphiQL UX may share similar challenges with Storybook (addon architecture, styling, etc).  Happy to share sketch files for [Storybook 5.0](https://medium.com/storybookjs/meet-storybook-5-0s-new-design-fb5735bc2a81) or any other design resources I have.

**Comment by connorshea** [comment]
I'd really like the ability to supply GraphiQL with "example queries" that the user could pick from to test out my API. For example, if I have a GraphQL API for a library of video games, I could have an example query that lists all video games by Nintendo, or all video games for the Nintendo Switch. I think that'd make the API a lot easier for new users to pick up and play with.

**Comment by benjie** [comment]
Agreed, @connorshea, I've been thinking the same for demoing template queries. It could also be a way of surfacing persisted queries (if you don't have too many) - could be combined with a read-only query area (that still has the keyword hover functionality).

**Comment by orta** [comment]
Yeah agreed, I wireframed that case (because I'd like it too) because we also do persisted queries

**Comment by leebyron** [comment]
I'm re-establishing WG meetings for the rest of the year, let's bring this up in discussion at the next, but I agree it's worth spooling up it's own group of folks to stake a direction for the future of this critical tool

**Comment by acao** [comment]
@IvanGoncharov and I are discussing this particular working group, and this seemed like a good agenda for a GraphiQL WG meeting:

- how to organize/delegate work for GraphiQL WG
- who will review work for new contributors to GraphiQL
- should codemirror-graphql work be inclusive in the GaphiQL WG
- should we update Contributing.md and how?
- (potentially) triage existing issues, PRs
- (potentially) decide on what needs to be accomplished before beginning new UI design

**Comment by acao** [comment]
what you proposed when initiating this discussion @orta sounds good for a second WG meeting, but as someone who's made a PR against GraphiQL recently, we have some housekeeping/contributor recruiting to do before we can start brainstorming on a 2.0

**Comment by acao** [comment]
@AGS-  @asiandrummer, @wincent, any other past major contributors you can think of besides Tess to invite?

**Comment by hwillson** [comment]
@acao Apollo is definitely interested in being part of the GraphiQL WG, and will help with contributions. We're discussing who will be the lead rep for this on our side, and will post back shortly. Thanks!

**Comment by orta** [comment]
Hi all, yep (sorry I missed the last WG, it was the day before my last day at Artsy and I had to much to do ) 

I'm not in a rush, I only had low fidelity mockups and a few ideas - happy to pop by the WG and get a feel for what folks are after 👍

**Comment by hwillson** [comment]
@acao Would you mind including @JakeDawkins and @trevor-scheer as part of the working group? They'll help out from the Apollo side.

**Comment by acao** [comment]
@hwillson heck yeah! we would love that.

**Comment by tanaypratap** [comment]
I'm very much interested in being part of the GraphiQL WG. Please tell me what should I do?

**Comment by acao** [comment]
this is the place to be @tanaypratap!

So, I've chosen [next tuesday, 5/14/19 at 16:00:00 GMT](https://www.timeanddate.com/worldclock/meetingdetails.html?year=2019&month=5&day=14&hour=16&min=0&sec=0&p1=224&p2=179&p3=136&p4=37&p5=239&p6=101&p7=152) for the first meeting.  If that's too short of notice, we can wait til 5/21, but it seems like folks are eager to get started! Very exciting, this is what we were hoping for.

I will create a PR shortly to create an agenda for this new working group's first meeting, and then folks can make PRs to add themselves to the agenda as attendees (similar to how the main GraphQL WG is organized).

Despite the original intent of this PR, it seems like we will need multiple meetings? Unless folks feel otherwise? Whether this will be an ad-hoc WG or an ongoing one will remain for the group to decide.

@leebyron for this repository, would a pattern where I create`<root>/subgroups/graphiql/agendas` and `<root>/subgroups/graphiql/notes` folders make sense?

**Comment by FluorescentHallucinogen** [comment]
@orta 
> Could we find a way to integrate ( or support the integration of) the features from GraphQL playground and/or explorer?

As you all remember, GraphQL Playground uses components of GraphiQL under the hood, but was created to fill in the gaps in the GraphiQL features. The main thing is to find the line between adding everything to the GraphiQL itself and implementing all this in the form of plug-ins for GraphiQL. As @sgrove said it's very important to make GraphiQL as much pluggable as possible. E.g. support of GraphQL subscriptions and GraphQL live, defer, stream queries should be supported by GraphiQL itself, because they can't be added by plug-ins (please correct me if I'm wrong) while e.g. [GraphiQL Explorer](https://github.com/OneGraph/graphiql-explorer) or [GraphiQL Code Exporter](https://github.com/OneGraph/graphiql-code-exporter) already implemented as plug-ins for GraphiQL and I see no reason to embed them into the GraphiQL itself.

Would you mind including @imolorhe (and maybe me i.e. @FluorescentHallucinogen) as part of the working group? @imolorhe is the creator of Altair (https://github.com/imolorhe/altair, https://altair.sirmuel.design), a more powerful alternative to Prisma Playground and GraphiQL.

Altair has many features that I want to see in the GraphiQL such as:

* [Upload files via GraphQL](https://github.com/imolorhe/altair#upload-files-via-graphql): you can use binary files as variables to upload them via GraphQL to server (if it implemented the GraphQL multipart request specification).

* [Add queries and fragments from documentation](https://github.com/imolorhe/altair#add-queries-and-fragments-from-documentation): You can easily add a query (or fragment) you see in the documentation into the editor without having to manually type every field within the query.

The recent thing that excites me is fields autocompletion. This is a real pain to add fileds only one by one...

Do you know that [GraphiQL has the autocomplete feature](https://twitter.com/alexey_rodionov/status/1106620753850626050) for years, but due to bad UX the users didn't know about it?

Altair has an ["Insert all fields"](https://github.com/imolorhe/altair/issues/690) feature by `Ctrl+Shift+Enter` hotkey and an `"addQueryDepthLimit"` option to control the default autocompletion depth limit.

Also I'm working on proposals to improve autocompletion (insert all fields of type) in GraphiQL Explorer. See mockups and explanations here: https://github.com/OneGraph/graphiql-explorer/issues/10. @sgrove PTAL, I want to hear your feedback finally! :)

Also it would be nice to have features that improve DX, e.g. refactor/transform queries ("Extract selected fields to fragment" and "Extract input fields to variable"). See explanations here: https://github.com/imolorhe/altair/issues/745.

**Comment by FluorescentHallucinogen** [comment]
Would you mind including guys (@rohit-ravikoti) from @Novvum as part of the working group? They create amazing things for GraphQL Playground and the whole GraphQL ecosystem such as [Schema Polling for Playground](https://www.novvum.io/post/graphql-playground-v1-8-8-introducing-schema-polling), [SDL Explorer for Playground](https://www.novvum.io/post/the-new-sdl-view-in-graphql-playground-v1-8-5) and [GraphQL Birdseye](https://www.novvum.io/post/introducing-graphql-birdseye) (see https://github.com/prisma/graphql-playground/pull/967 for integration of GraphQL Birdseye into GraphQL Playground).

**Comment by acao** [comment]
See above for instructions on joining the working group meetings, and when the first meeting is. All of these folks are who we had in mind. Anyone who is interested in becoming a long term contributor to `graphiql` is welcome to join the meetings.

As mentioned above, we have at least one meeting just for housekeeping and establishing the working group before we can jump into formally planning a new version of the graphiql project. For the time being, feel free to brainstorm the next version in this thread, it's all very exciting.

At this point, we are just looking to recruit long term contributors to maintain the project. As @IvanGoncharov pointed out in the last GraphQL Working Group meeting, we need to make sure that the project is in good shape to start adding new features, etc.

Possibly I should start another thread for establishing this working group, so that this thread can remain focused on brainstorming new features?

**Comment by acao** [comment]
Ok folks! Here's a schedule of working group meetings. I've tried to add y'all to the attendees list. If you are missing, please add yourselves to the relevant google docs. These docs might become a PR against this repository soon. Until these meetings, please keep the WG discussion going here in this issue thread.

- [GraphiQL WG Meeting 1 - May 14th, 2019](https://docs.google.com/document/d/1bqkjd47ikBQjcNE895Azla8b1n6zI5GNFg7WyYugbBg/edit?usp=sharing): Repository Housekeeping/Formalizing the (potentially temporary) working group. This is a great meeting for folks who plan on being long term contributors to `graphiql`.
- [GraphiQL WG Meeting 2 - May 16th, 2019](https://docs.google.com/document/d/1ahVPWSo5Q8VVsCf_8yzSlE7wV2NF7N30GCLhGhz_tZY/edit?usp=sharing): Informal discussion of GraphiQL V1, no formal decisions made. This would be good for folks who want to be long term contributors _and_ stakeholders who will not be contributing as much directly to `graphiql` itself. Make suggestions as to how we can architect things better for your particular dependent integration here.
- [GraphiQL WG Meeting 3 - June 11th, 2019](https://docs.google.com/document/d/1YPZyjFJpM4tqN2eQ5o4KPGfl9Ah0YAZkt0yE8NvEIWo/edit?usp=sharing): Discussion of GraphiQL 1.0, formal decisions made based on previous discussions in WG and in github issue(s).

To be honest, I don't think we anticipated such a flurry of interest so quickly, so we are trying to keep up! Excited to see what comes out of all of this. Love all the work y'all are doing. Feel free to email me at rikki.schulte@gmail.com if you have any questions. I'm essentially just acting as a facilitator and I haven't even formally contributed to GraphiQL yet!

**Comment by rohit-ravikoti** [comment]
> Would you mind including guys (@rohit-ravikoti) from @Novvum as part of the working group? They create amazing things for GraphQL Playground and the whole GraphQL ecosystem such as [Schema Polling for Playground](https://www.novvum.io/post/graphql-playground-v1-8-8-introducing-schema-polling), [SDL Explorer for Playground](https://www.novvum.io/post/the-new-sdl-view-in-graphql-playground-v1-8-5) and [GraphQL Birdseye](https://www.novvum.io/post/introducing-graphql-birdseye) (see [prisma/graphql-playground#967](https://github.com/prisma/graphql-playground/pull/967) for integration of GraphQL Birdseye into GraphQL Playground).

I'd love to be a part of the group 😁.

**Comment by acao** [comment]
@rohit-ravikoti  see above comment!

**Comment by asiandrummer** [comment]
Hi everyone - sorry I'm late to the party. Really good to see this working group forming :D When Lee and I open-sourced GraphiQL library I definitely didn't foresee this much love for this piece of software, so I'm definitely grateful that you guys are excited to take care of this tool.

As you probably have noticed, I moved away from contributing to open source softwares for a while, and I'm in the middle of many personal matters to spare any personal time on active developments, but I thought perhaps my experience on working the tool could be useful for a little while at least. I believe I can also represent the needs of Facebook developers and share the feature requests and pain points we've come across so far. @acao - I'll go ahead and join the first meeting.

**Comment by acao** [comment]
@asiandrummer thanks for joining us despite everything! It'll be a blessing to all of us to have your insight tomorrow.

Here's the agenda for the meeting tomorrow:
https://github.com/graphql/graphiql/blob/master/wg-agendas/05-14-2019.md

**Comment by acao** [comment]
Hey folks! So, if you check the agenda doc above, we updated with the outcomes of this discussion. Most importantly, Facebook officially handed off GraphiQL and codemirror-graphql to the GraphQL Foundation in that WG call.

Here is where the GraphiQL 1.0 plugin interface (an outcome of the last WG) discussion has moved to:
https://github.com/graphql/graphiql/issues/829

Please chime in there if you can, or create other GraphiQL 1.0 related discussion issues on the graphiql project itself if you like.

We will be addressing the future of `codemirror-graphql` as part of this WG as well.

We are well underway working towards a monorepo where `graphiql`, `graphql-language-service` and `codemirror-graphql` will all reside.

I will be scheduling another WG meeting for June here before the week is out.

We are still looking for maintainers for the IDE ecosystem. If I've reached out to you and you're interested, you'll likely hear from me again soon.

**Comment by BrunoScheufler** [comment]
@acao this sounds great already, will share my thoughts in the issue. I'd also offer to contribute to the new plugin system and future graphiql development in its entirety, let me know if there's anything apart from the linked discussion regarding future features of graphiql and the ecosystem 👍

**Comment by tanaypratap** [comment]
Rikki,

I missed the initial discussion due to some work commitments. Is it
possible that I get included in the next one? If yes, where should I add my
name?

Thanks,
Tanay

**Comment by acao** [comment]
I'll add the agenda document shortly!

**Comment by acao** [comment]
Next meeting is slated for 5/18/19, same time as GraphQL WG:
https://github.com/graphql/graphiql/pull/840
Let me know if there are any major conflicts with this date (yes, short notice), otherwise I will merge this in a few hours

**Comment by benjie** [comment]
@acao It's June now, my friend 😁

**Comment by acao** [comment]
@benjie  6/18, my bad! haha

**Comment by benjie** [comment]
(Also the PR say May in the text.) I can make Tuesday June 18th

**Comment by acao** [comment]
Tomorrow everyone! 16:00 UTC! :)
https://github.com/graphql/graphiql/blob/master/wg-agendas/06-18-2019.md

**Comment by tanaypratap** [comment]
thanks for the reminder Rikki! :)

**Comment by IvanGoncharov** [comment]
@acao I think this issue already solved. Can I close it?
BTW. I think it would be great to mention GraphiQL WG somewhere at the top of this repo README.

**Comment by acao** [comment]
yes you can close it! we are hoping for our own WG repo but thats another matter
